### ISSUE TITLE
[Feat?] make slide masters/layouts editable.

### ISSUE THREAD

**Issue by MP70** [singerla/pptx-automizer#85]
I think this would be 'not too difficult' to add; we just need to work out how to present it to the user.. maybe slide.modifyMaster()?

This would enable us to change text or logos that are stored in the masters. For example if a user puts the customer name and logo in the master we can still automate that without requiring them to pull them out of the master.

What do you think?

Alongside this I'd also like to work on getting slide backgrounds something we can edit.

**Comment by singerla** [comment]
Hi Matt,

I really like your idea, but I'm afraid, this is more complicated than expected. :) 

It's basically because a modification applies only on *adding* it - regardless of a shape's location (slide or slideMaster). It can't be modified *in place*, this would require some more refactoring.

But I could imagine a shurtcut like this:
```ts
 const pres = await automizer
    .loadRoot(`EmptyTemplate.pptx`)
    .load('SlideWithShapes.pptx')

    // modifyMaster could wrap around addMaster:
    .modifyMaster('SlideWithShapes.pptx', 1, (master) => {
      master.modifyElement(
        `MasterRectangle`,
        ModifyTextHelper.setText('my text on master'),
      );
    })
```
This is similar to `autoImportSlideMasters: true`, which also is a wrapper for underlying fetaures.
I need to go deeper into this.

**Comment by MP70** [comment]
Yes, I agree. I was thinking something that wrapped add master that 'offloaded' looking up the relationship and deduplication (in practice i suppose addMaster actually would overwrite the existing master). If we have one master applied to many slides it should ensure we are not actually copying the master each time and it remains a single master file.

Thinking about this further though there would potentially be cases where the master is on the root template, is there a way we could edit this once, without calling .addMaster on each slide manually.

**Comment by jhaase1** [comment]
I tried adding the masters making the modifications, then removing the originals. It does add the masters then remove the old ones; however, I am running into the repair issue. I believe the issue might be that the `id` for the masters might be too low if there are only modified masters in the presentation because without modification the slide master ids are 2147483660 to 2147483678, but after modification are 0 & 6.

If the default master id were increased, I think it may solve the issue?

https://github.com/singerla/pptx-automizer/blob/841b33ce0040e80cee6dab3b7ba11295a38b6c06/src/helper/modify-presentation-helper.ts#L65-L65

async function removeMasters(pres, length, from) {
  const xml = await XmlHelper.getXmlFromArchive(
    pres.rootTemplate.archive,
    `ppt/presentation.xml`,
  );

  const existingMasters = xml.getElementsByTagName('p:sldMasterId');

  XmlHelper.sliceCollection(existingMasters, length, from);

  XmlHelper.writeXmlToArchive(
    await pres.rootTemplate.archive,
    `ppt/presentation.xml`,
    xml,
  );

  await pres.modify(ModifyPresentationHelper.normalizeSlideMasterIds);
}
```

**Comment by jhaase1** [comment]
Big surprise it's more complicated than that. I don't think FileHelper is the correct function to use here? possibly remove from `pres.rootTemplate.archive.archive.files`? Or does it need to be removed from `pres.rootTemplate.archive.file` also?

```
async function removeMasters(pres, length, from) {
  const xml = await XmlHelper.getXmlFromArchive(
    pres.rootTemplate.archive,
    `ppt/presentation.xml`,
  );

  for (let i = 0; i < length; i += 1) {
    const layouts = await new XmlRelationshipHelper().initialize(
      pres.rootTemplate.archive,
      `slideMaster${from + i + 1}.xml.rels`,
      `ppt/slideMasters/_rels`,
      '../slideLayouts/slideLayout',
    );

    const layoutFiles = layouts.map((f) =>
      f.file, // path.resolve(`ppt/presentation.xml/${f.file}`),
    );

    await FileHelper.removeFromDirectory(
      pres.rootTemplate.archive,
      '../slideLayouts/',
      (file) => {
        return !layoutFiles.includes(file.relativePath);
      },
    );

    const themes = await new XmlRelationshipHelper().initialize(
      pres.rootTemplate.archive,
      `slideMaster${from + i + 1}.xml.rels`,
      `ppt/slideMasters/_rels`,
      '../theme/theme',
    );

    const themesFiles = themes.map((f) =>
      f.file, // path.resolve(`ppt/presentation.xml/${f.file}`),
    );

    await FileHelper.removeFromDirectory(
      pres.rootTemplate.archive,
      '../theme/',
      (file) => {
        return !themesFiles.includes(file.relativePath);
      },
    );
  }

  const existingMasters = xml.getElementsByTagName('p:sldMasterId');

  XmlHelper.sliceCollection(existingMasters, length, from);

  XmlHelper.writeXmlToArchive(
    await pres.rootTemplate.archive,
    `ppt/presentation.xml`,
    xml,
  );

  // await pres.modify(ModifyPresentationHelper.normalizeSlideMasterIds);
}

**Comment by singerla** [comment]
Hi @jhaase1!

Thanks a lot for your code. I've edited it a little and moved it to https://github.com/singerla/pptx-automizer/blob/794870529ecb992ce667bb094185b5770c91c3dc/src/helper/modify-presentation-helper.ts#L190.

Please check out `feature-remove-master` branch to test and improve it.

Some thoughts on it:
- It is really tricky to figure out a proper `Id` for the first master. Please also see my comments on https://github.com/singerla/pptx-automizer/blob/794870529ecb992ce667bb094185b5770c91c3dc/src/helper/modify-presentation-helper.ts#L46-L54
- I could not figure out why deleting the theme files causes a corrupted pptx on.
- This will eventually not remove e.g. images place on a remove slide master.

Please let me know if this is useful!

**Comment by jhaase1** [comment]
I manually removed a slide master!

Steps:

- Delete theme1
- Delete slideMaster1
- Delete slideMaster1.xml.rels
- Delete slideLayout1-8.xml
- Delete slideLayout1-8.xml.rels
- Presentation.xml.rels has one reference to theme1 -> I changed it to theme3
- Presentation.xml sldMasterIdLst reference to rId1 - delete

[original.pptx](https://github.com/user-attachments/files/18355692/original.pptx)

[master-removed.pptx](https://github.com/user-attachments/files/18355693/master-removed.pptx)

**Comment by singerla** [comment]
Cool, @jhaase1! 

Did you also try this:
https://github.com/singerla/pptx-automizer/blob/794870529ecb992ce667bb094185b5770c91c3dc/__tests__/add-slide-master.test.ts#L57

It is available in branch `feature-remove-master`. I was wondering if this is applicable to your case.

**Comment by jhaase1** [comment]
I thought you were saying the function was not working because there was something missing. That's why I was doing the manual editing to try and figure out what we needed to add.

I'll test this out!

**Comment by jhaase1** [comment]
Thank you! That's exactly what I was hoping for :-)

**Comment by jhaase1** [comment]
After looking move closely I realized that you had broken the test to get the POC working. I went back and changed things around to make the `removeMasters` match other functions such as `addSlide` by adding a return value. This fixed the error that `pres.masters` did not exist; however, I cannot figure out where that is actually being set because the value should be 2?

https://github.com/singerla/pptx-automizer/pull/142
